db/db_compaction_test.cc

```diff
@@ -6181,6 +6181,231 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
   Close();
 }
 
+class DBCompactionTestWithOngoingFileIngestionParam
+    : public DBCompactionTest,
+      public testing::WithParamInterface<std::string> {
+ public:
+  DBCompactionTestWithOngoingFileIngestionParam() : DBCompactionTest() {
+    compaction_path_to_test_ = GetParam();
+  }
+  void SetupOptions() {
+    options_ = CurrentOptions();
+    options_.create_if_missing = true;
+
+    if (compaction_path_to_test_ == "RefitLevelCompactRange") {
+      options_.num_levels = 7;
+    } else {
+      options_.num_levels = 3;
+    }
+    options_.compaction_style = CompactionStyle::kCompactionStyleLevel;
```

Comment:
Tests look great. I assume the issue also applies to other compaction styles?

Reply:
I doubt for other compactions. Assume the WaitForIngestFile() is already removed, in order for the issue happens, we need a compaction to output files to the output level of file ingestion. The compaction could be user-specified-level compaction (RefitLevel, CompactFiles(level)) or non-user-specified-level compaction (all other compactions.)

Reply:
> I assume the issue also applies to other compaction styles?

I doubt it for other compaction styles. 

For multi-level FIFO, the non-deletion compaction (the one has output files i.e,  FindIntraL0Compaction in PickSizeCompaction, PickCompactionToWarm, CompactFiles(output_level <= FIFOCompactionPicker::MaxOutputLevel())) are all L0->L0 compaction. So compaction output will stay in L0 where overlapping range is allowed as well as we won't have inverted-level for new/old data. 

For universal, it's a bit tricky. 
- For RefitLevel(change_level=true), the one our PR is motivated to fix, I believe universal has been safe. This is because https://github.com/facebook/rocksdb/blob/7.8.fb/db/external_sst_file_ingestion_job.cc#L799 leads to files ingested to non-L0 level only if that level is non-empty, under universal compaction (though I didn't spend time in fully understanding the intention). So the compaction output will output files to a non-empty level, which can't be the target_level of RefitLevel(change_level=true), guaranteed by existing codes.
- For non-RefitLevel() compaction, the one our PR replace the fixes to,  I believe it's possible for universal to hit the issues (without WaitForIngestFiles() nor our fix). It's just not easy to write a UT to ingest file to non-L0 for universal compaction due to https://github.com/facebook/rocksdb/blob/7.8.fb/db/external_sst_file_ingestion_job.cc#L828-L829 and I decided to trust the way we are replacing the previous fix is generic enough for universal compaction.